pandas/core/frame.py

```diff
@@ -8559,6 +8580,54 @@ def blk_func(values):
                 out[:] = coerce_to_dtypes(out.values, df.dtypes)
             return out
 
+        def array_func(values):
+            if isinstance(values, ExtensionArray):
+                return values._reduce(name, skipna=skipna, **kwds)
+            else:
+                return op(values, skipna=skipna, **kwds)
+
+        # all other options with axis=0 are done column-array-wise
+        if axis == 0:
+
+            def _constructor(df, result, index=None):
+                index = index if index is not None else df.columns
+                if len(result):
+                    return df._constructor_sliced(result, index=index)
+                else:
+                    # set correct dtype for empty result
+                    dtype = "bool" if filter_type == "bool" else "float64"
+                    return df._constructor_sliced(result, index=index, dtype=dtype)
+
+            def _reduce_columns(df, op):
+                result = [op(arr) for arr in df._iter_column_arrays()]
+                return _constructor(df, result)
+
+            df = self
+            if numeric_only is True:
+                df = _get_data(axis_matters=True)
+
+            if numeric_only is not None:
+                return _reduce_columns(df, array_func)
+            else:
+                # need to catch and ignore exceptions when numeric_only=None
+                try:
+                    return _reduce_columns(df, array_func)
+                except TypeError:
+                    # if column-wise fails and numeric_only was None, we try
```

Comment:
is the try/except here needed?  why not go straight to the except?

Reply:
Ah, good point. I think I originally did this in the idea that "this is faster for the cases when no exceptions would be raised". But I suppose the cost of also keeping track of the `indices` (which is the main difference with `_reduce_columns`) isn't very big. Will check.

In principle also the `if numeric_only is True:` one could be combined, if we move the `numeric_only` check inside the try/except in the inner loop

Reply:
> In principle also the if numeric_only is True: one could be combined, if we move the numeric_only check inside the try/except in the inner loop

Have to admit I'm having trouble following all the cases.  Have you already split off pieces that can be addressed independently? (e.g. you mentioned td64/dt64 .all() and some empty-frame behavior)

Reply:
Removed the try/except here, directly use the code that was now in the except clause